lib/HeroBonus.cpp

```diff
@@ -452,30 +455,41 @@ void BonusList::stackBonuses()
 
 int BonusList::totalValue() const
 {
-	int base = 0;
+	//BASE_NUMBER, PERCENT_TO_BASE, PERCENT_TO_ALL, ADDITIVE, PERCENT_TO_SOURCE.
+	//PERCENT_TO_TARGET_TYPE is a marker here, it should always be after percent to source
+	//Bonuses before should be modified by source, bonuses after - not
+	using BonusCollection = std::array<int, Bonus::PERCENT_TO_TARGET_TYPE>;

```

Comment:
Still thinking that using struct instead of array would be more clear and won't require huge comment to explain the code.

Reply:
@IvanSavenko but 3 lines of comment is less LOC than entire new struct which will function as array. I really do not like struct here.

Reply:
3 lines of *code* and 3 lines of *comments* have very different cognitive load.

Compare
```c++
	//BASE_NUMBER, PERCENT_TO_BASE, PERCENT_TO_ALL, ADDITIVE, PERCENT_TO_SOURCE.
	//PERCENT_TO_TARGET_TYPE is a marker here, it should always be after percent to source
	//Bonuses before should be modified by source, bonuses after - not
	using BonusCollection = std::array<int, Bonus::PERCENT_TO_TARGET_TYPE>;
```

```c++
struct BonusCollection
{
    int baseNumber;
    int percentToBase;
    int percentToAll;
    int additive;
    int percentToSource
};
```
Yes, struct version is (oh, noes) 6 lines of code + 2 empty lines, but time to understand what this code does would be much shorter for struct.

Besides - you're not using this array as well, array.
This method has 0 loops over this array and 0 accesses by (non-constant) indexes.
The only access to this array is by compile-time constant index. Which is what struct does.

`sources[b->source].at(Bonus::BASE_NUMBER) += b->val;`
is equivalent to
`sources[b->source].baseNumber += b->val;`

`modifiedBase += src.at(Bonus::BASE_NUMBER) + (src.at(Bonus::BASE_NUMBER) * src.at(Bonus::PERCENT_TO_SOURCE)) / 100;`
is equivalent to
`modifiedBase += src.baseNumber + (src.baseNumber * src.percentToSource) / 100;`

And in both cases, expression that uses struct is less complex.

Reply:
@IvanSavenko Ok, rewrote as struct by your opinion.